File: Android/app/src/main/java/com/swarnkary/donezy/Achievements.kt
package com.swarnkary.donezy

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.WeekFields
import java.util.Locale

/**
 * Achievements are derived from the existing log/hobby data — no extra table.
 * Each achievement has a stable id (used in undo/celebration text) and a predicate that
 * answers: "given this snapshot, has the user earned this badge?".
 *
 * To detect *newly* earned achievements after a save, the ViewModel runs the same set of
 * predicates against `before` and `after` snapshots and emits whatever transitioned to true.
 */
data class Achievement(
    val id: String,
    val emoji: String,
    val title: String,
    val tagline: String
)

object Achievements {

    val ALL: List<Pair<Achievement, AchievementSnapshot.() -> Boolean>> = listOf(
        Achievement("first_log",     "✨", "First log",       "You've started — the rest is just keeping going.")
            to { totalLogs >= 1 },
        Achievement("five_logs",     "🌱", "5 logs",          "Habit forming — momentum is real.")
            to { totalLogs >= 5 },
        Achievement("twenty_five",   "🌿", "25 logs",         "Roots are deep. This tracker is part of you now.")
            to { totalLogs >= 25 },
        Achievement("hundred",       "🌳", "100 logs",        "A century of check-ins — incredible commitment.")
            to { totalLogs >= 100 },
        Achievement("first_photo",   "📸", "First photo",     "Memory captured. Future-you will thank you.")
            to { hasPhoto },
        Achievement("five_star",     "⭐", "5-star session",  "A perfect session deserves to be remembered.")
            to { fiveStar },
        Achievement("daily_streak",  "🔥", "Streak started",  "Day 1. Show up tomorrow and it doubles.")
            to { currentStreak >= 1 },
        Achievement("three_streak",  "🔥", "3-day streak",    "Three days in a row — the habit's taking hold.")
            to { currentStreak >= 3 },
        Achievement("week_streak",   "🔥", "7-day streak",    "A whole week. The hard part is over.")
            to { currentStreak >= 7 },
        Achievement("two_weeks",     "🔥", "14-day streak",   "Two weeks running — this is who you are now.")
            to { currentStreak >= 14 },
        Achievement("month_streak",  "🏆", "30-day streak",   "A month-long streak. Take a moment to be proud.")
            to { currentStreak >= 30 },
        Achievement("hundred_streak","👑", "100-day streak",  "Triple digits. You're an inspiration.")
            to { currentStreak >= 100 },
        Achievement("weekly_goal",   "🎯", "Weekly goal hit", "You hit this week's target. Treat yourself.")
            to { weeklyGoal in 1..logsThisWeek },
        Achievement("comeback",      "💫", "Comeback",        "Welcome back — picking it up again is the bravest part.")
            to { comeback }
    )

    /** Predicates that turned `false → true` between two snapshots. */
    fun newlyEarned(before: AchievementSnapshot, after: AchievementSnapshot): List<Achievement> =
        ALL.filter { (_, pred) -> !before.pred() && after.pred() }.map { it.first }

    /** All currently-earned achievements for a tracker, used by detail screen. */
    fun earned(snapshot: AchievementSnapshot): List<Achievement> =
        ALL.filter { (_, pred) -> snapshot.pred() }.map { it.first }
}

/**
 * Pre-computed numbers fed into achievement predicates. Keeping it small keeps the
 * predicate set easy to extend.
 */
data class AchievementSnapshot(
    val totalLogs: Int,
    val currentStreak: Int,
    val hasPhoto: Boolean,
    val fiveStar: Boolean,
    val logsThisWeek: Int,
    val weeklyGoal: Int,
    /** True when the user logs today after a gap of 7+ days since the previous log. */
    val comeback: Boolean
) {
    companion object {
        fun from(hobby: Hobby, logs: List<HobbyLog>, zone: ZoneId = ZoneId.systemDefault()): AchievementSnapshot {
            val streak = HobbyViewModel.computeStreak(logs)
            val today = LocalDate.now(zone)
            val weekFields = WeekFields.of(Locale.getDefault())
            val thisWeek = today.get(weekFields.weekOfWeekBasedYear())
            val thisWeekYear = today.get(weekFields.weekBasedYear())
            val logsThisWeek = logs.count {
                val d = Instant.ofEpochMilli(it.createdAt).atZone(zone).toLocalDate()
                d.get(weekFields.weekOfWeekBasedYear()) == thisWeek &&
                    d.get(weekFields.weekBasedYear()) == thisWeekYear
            }

            // Comeback: the most recent two logs are >=7 days apart, and the latest is today.
            val sorted = logs.sortedByDescending { it.createdAt }
            val comeback = if (sorted.size >= 2) {
                val newest = Instant.ofEpochMilli(sorted[0].createdAt).atZone(zone).toLocalDate()
                val prev = Instant.ofEpochMilli(sorted[1].createdAt).atZone(zone).toLocalDate()
                newest == today && java.time.temporal.ChronoUnit.DAYS.between(prev, newest) >= 7
            } else false

            return AchievementSnapshot(
                totalLogs = logs.size,
                currentStreak = streak,
                hasPhoto = logs.any { it.photoUri != null },
                fiveStar = logs.any { it.rating == 5 },
                logsThisWeek = logsThisWeek,
                weeklyGoal = hobby.weeklyGoal,
                comeback = comeback
            )
        }
    }
}

/** A small bank of varied affirmations so the same toast doesn't repeat after every save. */
object Affirmations {
    private val random = java.util.Random()

    private val PHRASES = listOf(
        "Log saved ✓ · Nice work",
        "Logged · Small steps, big wins",
        "Saved · Future-you will thank present-you",
        "Logged · One more brick in the wall",
        "Saved · That's another rep for the streak",
        "Logged · You showed up today",
        "Saved · Quiet consistency is the move",
        "Logged · Progress, not perfection",
        "Saved · Keep stacking days",
        "Logged · This is how habits stick"
    )

    fun pick(): String = PHRASES[random.nextInt(PHRASES.size)]
}

File: MacOs/Shared/Achievements.swift
import Foundation

/// Achievements are derived from existing log/hobby data — no extra table.
/// Each has a stable id and a predicate answering: "given this snapshot, earned?".
struct Achievement: Identifiable, Equatable {
    let id: String
    let emoji: String
    let title: String
    let tagline: String
}

enum Achievements {

    /// (achievement, predicate) pairs, in display order. Mirrors the Android `ALL` list.
    static let all: [(Achievement, (AchievementSnapshot) -> Bool)] = [
        (Achievement(id: "first_log", emoji: "✨", title: "First log", tagline: "You've started — the rest is just keeping going."),
            { $0.totalLogs >= 1 }),
        (Achievement(id: "five_logs", emoji: "🌱", title: "5 logs", tagline: "Habit forming — momentum is real."),
            { $0.totalLogs >= 5 }),
        (Achievement(id: "twenty_five", emoji: "🌿", title: "25 logs", tagline: "Roots are deep. This tracker is part of you now."),
            { $0.totalLogs >= 25 }),
        (Achievement(id: "hundred", emoji: "🌳", title: "100 logs", tagline: "A century of check-ins — incredible commitment."),
            { $0.totalLogs >= 100 }),
        (Achievement(id: "first_photo", emoji: "📸", title: "First photo", tagline: "Memory captured. Future-you will thank you."),
            { $0.hasPhoto }),
        (Achievement(id: "five_star", emoji: "⭐", title: "5-star session", tagline: "A perfect session deserves to be remembered."),
            { $0.fiveStar }),
        (Achievement(id: "daily_streak", emoji: "🔥", title: "Streak started", tagline: "Day 1. Show up tomorrow and it doubles."),
            { $0.currentStreak >= 1 }),
        (Achievement(id: "three_streak", emoji: "🔥", title: "3-day streak", tagline: "Three days in a row — the habit's taking hold."),
            { $0.currentStreak >= 3 }),
        (Achievement(id: "week_streak", emoji: "🔥", title: "7-day streak", tagline: "A whole week. The hard part is over."),
            { $0.currentStreak >= 7 }),
        (Achievement(id: "two_weeks", emoji: "🔥", title: "14-day streak", tagline: "Two weeks running — this is who you are now."),
            { $0.currentStreak >= 14 }),
        (Achievement(id: "month_streak", emoji: "🏆", title: "30-day streak", tagline: "A month-long streak. Take a moment to be proud."),
            { $0.currentStreak >= 30 }),
        (Achievement(id: "hundred_streak", emoji: "👑", title: "100-day streak", tagline: "Triple digits. You're an inspiration."),
            { $0.currentStreak >= 100 }),
        (Achievement(id: "weekly_goal", emoji: "🎯", title: "Weekly goal hit", tagline: "You hit this week's target. Treat yourself."),
            { $0.weeklyGoal >= 1 && $0.weeklyGoal <= $0.logsThisWeek }),
        (Achievement(id: "comeback", emoji: "💫", title: "Comeback", tagline: "Welcome back — picking it up again is the bravest part."),
            { $0.comeback }),
    ]

    /// Predicates that turned false → true between two snapshots.
    static func newlyEarned(before: AchievementSnapshot, after: AchievementSnapshot) -> [Achievement] {
        all.filter { !$0.1(before) && $0.1(after) }.map { $0.0 }
    }

    /// All currently-earned achievements for a tracker.
    static func earned(_ snapshot: AchievementSnapshot) -> [Achievement] {
        all.filter { $0.1(snapshot) }.map { $0.0 }
    }
}

/// Pre-computed numbers fed into achievement predicates.
struct AchievementSnapshot: Equatable {
    let totalLogs: Int
    let currentStreak: Int
    let hasPhoto: Bool
    let fiveStar: Bool
    let logsThisWeek: Int
    let weeklyGoal: Int
    /// True when the user logs today after a gap of 7+ days since the previous log.
    let comeback: Bool

    static func from(hobby: Hobby, logs: [HobbyLog]) -> AchievementSnapshot {
        let calendar = Calendar.current
        let streak = StreakMath.computeStreak(logs)
        let today = calendar.startOfDay(for: Date())

        func weekKey(_ date: Date) -> Int {
            let comps = calendar.dateComponents([.weekOfYear, .yearForWeekOfYear], from: date)
            return (comps.yearForWeekOfYear ?? 0) * 100 + (comps.weekOfYear ?? 0)
        }
        let thisWeekKey = weekKey(today)
        let logsThisWeek = logs.filter {
            weekKey(Date(timeIntervalSince1970: Double($0.createdAt) / 1000.0)) == thisWeekKey
        }.count

        // Comeback: the most recent two logs are >=7 days apart, and the latest is today.
        let sorted = logs.sorted { $0.createdAt > $1.createdAt }
        var comeback = false
        if sorted.count >= 2 {
            let newest = calendar.startOfDay(for: Date(timeIntervalSince1970: Double(sorted[0].createdAt) / 1000.0))
            let prev = calendar.startOfDay(for: Date(timeIntervalSince1970: Double(sorted[1].createdAt) / 1000.0))
            let gap = calendar.dateComponents([.day], from: prev, to: newest).day ?? 0
            comeback = newest == today && gap >= 7
        }

        return AchievementSnapshot(
            totalLogs: logs.count,
            currentStreak: streak,
            hasPhoto: logs.contains { $0.photoUri != nil },
            fiveStar: logs.contains { $0.rating == 5 },
            logsThisWeek: logsThisWeek,
            weeklyGoal: hobby.weeklyGoal,
            comeback: comeback
        )
    }
}

/// A small bank of varied affirmations so the same toast doesn't repeat every save.
enum Affirmations {
    private static let phrases = [
        "Log saved ✓ · Nice work",
        "Logged · Small steps, big wins",
        "Saved · Future-you will thank present-you",
        "Logged · One more brick in the wall",
        "Saved · That's another rep for the streak",
        "Logged · You showed up today",
        "Saved · Quiet consistency is the move",
        "Logged · Progress, not perfection",
        "Saved · Keep stacking days",
        "Logged · This is how habits stick",
    ]

    static func pick() -> String { phrases.randomElement() ?? phrases[0] }
}
